### ISSUE TITLE
Understanding dependent rules

### ISSUE THREAD

**Issue by infinitimods** [JeremySkinner/FluentValidation#153]
(This has been moved here from CodePlex)

Dear Jeremy,

I have been reading about dependent rules here:

https://github.com/JeremySkinner/FluentValidation/issues/65

From my understanding, a dependent rule will execute only if the preceding rules succeed. However, I cannot get this to work. Ex:

```
    internal class CustomerValidator : AbstractValidator<Customer>
    {
        internal CustomerValidator()
        {
            RuleSet("Add", () =>
            {
                RuleFor(o => o).NotNull()
                    .DependentRules(d =>
                    {
                        d.RuleFor(x => x.FirstName).NotEmpty();
                    });

                //this.BaseValidation();              
            });
        }
    }
```

If I pass in a customer object which is not null, I expect to validate the first name. However, no validation is fired. Upon further investigation, it appears that DependentRules do not work inside the RuleSet. Is it possible to make it work inside the RuleSet?

Thank you for your help,

Michael

**Comment by infinitimods** [comment]
(This is a reply from Jeremy)

Hi Michael,

FluentValidation has moved to GitHub - in future please post your questions over there rather than on codeplex: https://github.com/jeremyskinner/fluentvalidation/issues



So at the moment Dependent Rules aren't compatible with rulesets - I'll get this fixed. 

(Additionally, FluentValidation requires that the root-level object being validated is not-null, so your example using RuleFor(o => o).NotNull() won't actually work, but the concept is right)

**Comment by infinitimods** [comment]
Hi Jeremy,

Thank you for a swift reply. It would also be super useful for DependentRules to execute any code in inside that code block, not just rules. I've tried doing this yesterday, but it also did not work:

    internal class CustomerValidator : AbstractValidator<Customer>
    {
        /// <summary>
        /// 
        /// </summary>
        internal CustomerValidator()
        {
			RuleFor(o => o)
				.NotNull().WithMessage("Customer request cannot be null")
				.DependentRules(o =>
				{
					RuleSet(RuleSets.Add, () =>
					{
						this.BaseValidation();
					});

					RuleFor(x => x.Id).NotEmpty().WithMessage("Customer id is required")
						.Must(BeAValidCustomer).WithMessage("This customer has been removed or does not exist. Please try navigating to this page again")
						.DependentRules(x =>
						{
							RuleSet(RuleSets.Update, () =>
							{
								this.BaseValidation();
							});

							RuleSet(RuleSets.Delete, () =>
							{
								this.BaseValidation();
							});
						});
				});
        }
    }

**Comment by JeremySkinner** [comment]
So I've made the change in master that will enable support for rulesets.

Note that you can't put just any code inside a call to DependentRules, you can only invoke methods on the parameter itself. So this will work:

```
 .DependentRules(d =>                   {
                        d.RuleSet("Update", () =>
                        {
                            d.RuleFor();...
                            d.RuleFor();
                        });
                    });
```

But this will not:

```
 .DependentRules(d =>                   {
                        RuleSet("Update", () =>
                        {
                            RuleFor();...
                           RuleFor();
                        });
                    });
```

You must invoke the methods on the "d" parameter - otherwise there's no way for FV to know that the rule is dependent.

**Comment by infinitimods** [comment]
Jeremy, thank you for support! I haven't been expecting a turned around this quickly!

Is it possible to add support for a method call inside the RuleSet that is encapsulated by DependentRules? I am trying to avoid code duplication and my BaseValidation method holds all the rest of the rules needed for object validation.

**Comment by JeremySkinner** [comment]
No, sorry, the DependentRules implementation doesn't work that way - you can't just make normal method calls inside here- you must call methods on the parameter as I mentioned above.

What you can do though, is modify your BaseValidation method to take a parameter which you invoke rules on. Example:

```
public class CustomerValidator : AbstractValidator<Customer>
{
    public CustomerValidator()
    {
    RuleFor(...).Whatever().
       .DependentRules(d =>                   {
                        d.RuleSet("Update", () =>
                        {
                            BaseValidation(d);
                        });
                    });
    }
    
    public void BaseValidation(AbstractValidator<Customer> validator) {
     // Call v.RuleFor rather than just RuleFor
      validator.RuleFor(x => x.Foo).Whatever(); 
    }
}
```

**Comment by infinitimods** [comment]
Jeremy, that should work! Let me give it a try! One small bug that I found, when I do this:

			RuleSet("Add", () =>
			{
				RuleFor(o => o).NotNull()
					.DependentRules(d =>
					{
						d.RuleFor(x => x.FirstName).NotEmpty().WithMessage("First name cannot be empty");
					});      
			});

It works as expected, but the validation error that I am getting is: " -- '' must not be empty " and it's ignoring my custom message.

**Comment by infinitimods** [comment]
Hmm, weird. It is working now. Must have been something on my end. I'll be testing more thoroughly today. Jeremy, thank you very much for the support!!

**Comment by infinitimods** [comment]
Hi Jeremy,

Trying to understand as why my BaseValidation() method will not fire off in the code below assuming first name is just an empty string. Is there a workaround for that?

	internal class CustomerValidator : AbstractValidator<Customer>
	{
		internal CustomerValidator()
		{

			RuleSet(RuleSets.Add, () =>
			{
				RuleFor(o => o)
					.NotNull().WithMessage("Customer cannot be null")
					.DependentRules(o =>
					{
						this.BaseValidation(o);
					});
			});

			RuleSet(RuleSets.Update, () =>
			{
				RuleFor(o => o)
					.NotNull().WithMessage("Customer cannot be null on update")
					.DependentRules(o =>
					{
						o.RuleFor(x => x.Id)
							.Cascade(CascadeMode.StopOnFirstFailure)
							.NotEmpty().WithMessage("Customer id is required")
							.Must(BeAValidCustomer).WithMessage("This customer has been removed")
							.DependentRules(y =>
							{
								this.BaseValidation(y);
							});
					});
			});

		}

		internal void BaseValidation(AbstractValidator<Customer> validator)
		{
			validator.RuleFor(o => o.FirstName)
				.Cascade(CascadeMode.StopOnFirstFailure)
				.NotEmpty().WithMessage("First name is required");
		}
	}

**Comment by infinitimods** [comment]
I've also tried to cascade dependent rules directly. Everything works fine up the first name validation, but last name validation never fires off.

				RuleFor(x => x.Id)
					.Cascade(CascadeMode.StopOnFirstFailure)
					.NotEmpty().WithMessage("Customer id is required")
					.Must(BeAValidCustomer).WithMessage("This customer has been removed")
							.DependentRules(y =>
							{
								y.RuleFor(z => z.FirstName).NotEmpty().WithMessage("First name is required")
										.DependentRules(a =>
										{
											a.RuleFor(b => b.LastName).NotEmpty().WithMessage("Last name is required");
										});
							});
				}

**Comment by JeremySkinner** [comment]
I don't think nested dependent rules will work properly - it was really only designed for going 1 level down. 

If you could put together a unit test that illustrates this, then I'll have a play with it and see how much work this'd be to implement.

**Comment by infinitimods** [comment]
Hi Jeremy, here are couple of unit tests that I wrote and tested them out in Visual Studio. Both of them are essentially the same and fail currently as they should. Nested dependent rules will be awesome to have. Unfortunately I do not understand fluent framework and a lot of hardcore c# syntax that's used well enough to do this on my own at this point.

		[Fact]
		public void Nested_dependent_rules_inside_ruleset()
		{
			var validator = new TestValidator();
			validator.RuleSet("MyRuleSet", () =>
			{

				validator.RuleFor(x => x.Surname).NotNull()
					.DependentRules(d =>
					{
						d.RuleFor(x => x.Forename).NotNull()
							.DependentRules(y =>
							{
								y.RuleFor(x => x.Address).NotNull();
							});
					});
			});

			var results = validator.Validate(new Person { Surname = "foo", Forename="foo" }, ruleSet: "MyRuleSet");
			results.Errors.Count.ShouldEqual(1);
			results.Errors.Single().PropertyName.ShouldEqual("Address");
		}

		[Fact]
		public void Nested_dependent_rules_inside_ruleset_inside_method()
		{
			var validator = new TestValidator();
			validator.RuleSet("MyRuleSet", () =>
			{

				validator.RuleFor(x => x.Surname).NotNull()
					.DependentRules(d =>
					{
						d.RuleFor(x => x.Forename).NotNull()
							.DependentRules(y =>
							{
								BaseValidation(y);
							});
					});
			});
			
			var results = validator.Validate(new Person { Surname = "foo", Forename = "foo" }, ruleSet: "MyRuleSet");
			results.Errors.Count.ShouldEqual(1);
			results.Errors.Single().PropertyName.ShouldEqual("Address");
		}

		private void BaseValidation(DependentRules<Person> validator)
		{
			validator.RuleFor(x => x.Address).NotNull();
		}

**Comment by infinitimods** [comment]
Hi Jeremy! Just wondering if you had a chance to play around with it and see if this is something that can be done without requiring too much effort. Thank you for all your help!

**Comment by JeremySkinner** [comment]
I haven't had a chance yet I'm afraid...I've got a pretty big backlog of FV issues to look at, I promise I will get to it though!

**Comment by infinitimods** [comment]
Thank you Jeremy! No worries, whenever you get a chance.

**Comment by Misiu** [comment]
@JeremySkinner did You had time to look at this? I'm building complex validation class and multiple DependentRules would help me a lot. For now I'm using `When` but this way I don't get correct error messages.

**Comment by JeremySkinner** [comment]
Sorry, no- I'm not currently doing any work on FluentValidation as I'm busy with other commitments, so unless someone wants to contribute a pull request I'm afraid this will have to wait.

**Comment by Misiu** [comment]
@JeremySkinner as I wrote I have a workaround for that. Sadly I'm not that advanced developer to change this.

**Comment by JeremySkinner** [comment]
This is actually implemented as of 7.4, but you must use the new/updated DependentRules syntax:

```
RuleSet("MyRuleSet", () => {
  RuleFor(x => x.Surname).NotNull().DependentRules(() => {
    RuleFor(x => x.Forename).NotNull().DependentRules(() =>	{
      RuleFor(x => x.Address).NotNull();
    });
  });
});
```

Note that I'm using a new overload of DependentRules that takes a simple action, rather than a DependentRulesBuilder. Inside the action, I now calling the root-level `RuleFor` method instead of calling `d.RuleFor`. 

This is the recommended way of using DependentRules going forward.
